File: tests/fixtures/extensions/arc_progress_bar.cpp
#include "arc_progress_bar.hpp"

/**
 * @brief Get color from ImGui style safely.
 *
 * @param color_id ImGui color ID.
 * @return ImColor Return the color, or default color if invalid ID.
 */
static inline ImColor _GetStyleColor(const ImGuiCol_ color_id)
{
    IM_ASSERT(color_id >= 0 && color_id < ImGuiCol_COUNT);

    if (color_id < 0 || color_id >= ImGuiCol_COUNT)
        return ImColor(DEFAULT_FOREGROUND_COLOR);    // Fallback

    return ImColor(ImGui::GetStyle().Colors[color_id]);
}

/**
 * @brief The actual drawing function for arc progress bar.
 *
 * @param size Diameter of the arc.
 * @param max_angle_factor Arc's maximum angle factor (0~1, 1 for 360 degrees).
 * @param percentage Current progress percentage (0~1).
 * @param thickness Arc line thickness.
 * @param pos Position of the widget (top-left corner).
 */
void _DrawArc(float size, float max_angle_factor, float percentage, float thickness, ImVec2 pos)
{
    ImDrawList *draw_list = ImGui::GetWindowDrawList();

    float x = pos.x, y = pos.y;    // Position

    constexpr float ONE_DIV_360f = 1.0f / 360.0f;   // Performance tweak

    float a_min_factor, a_max_factor, a_max_factor_100percentage;
    a_min_factor = -1.5f + ((360 - max_angle_factor) * ONE_DIV_360f);    // Angle PI*-1.5f resides in the bottom point of circle
    a_max_factor_100percentage = (a_min_factor + 1.0f) * -1.0f;       // Arc max factor on 100% percentage state

    float a_factor_delta = (a_max_factor_100percentage - a_min_factor) * (percentage * 0.01f);
    a_max_factor = a_min_factor + a_factor_delta;

    // Path for background arc (dimmed arc)
    draw_list->PathArcTo(ImVec2(x + size * 0.5f, y + size * 0.5f), size * 0.5f, 3.141592f * a_min_factor, 3.141592f * a_max_factor_100percentage);
    draw_list->PathStroke(_GetStyleColor(ImGuiCol_Button), ImDrawFlags_None, thickness);

    // Path for progress filling (highlighted arc)
    draw_list->PathArcTo(ImVec2(x + size * 0.5f, y + size * 0.5f), size * 0.5f, 3.141592f * a_min_factor, 3.141592f * a_max_factor);
    draw_list->PathStroke(_GetStyleColor(ImGuiCol_ButtonActive), ImDrawFlags_None, thickness);
}

void ImGuiExt::ProgressBarArc(float size, float max_angle_factor, float percentage, ImVec2 pos, float thickness)
{
    // By default, ImDrawList draws in absolute position relaitve to the canvas, NOT the window.
    // So we need to consider the window position if we want to draw something inside the window.

    ImVec2 window_pos = ImGui::GetWindowPos();    // Do not forget to get window's position.

    // And do not forget to add the window position to our own position.
    _DrawArc(size, max_angle_factor, percentage, thickness, ImVec2(pos.x + window_pos.x, pos.y + window_pos.y));
}

void ImGuiExt::ProgressBarArc(float size, float max_angle_factor, float percentage, float thickness)
{
    ImVec2 pos = ImGui::GetCursorScreenPos(); // Get current insert point
    ImGui::Dummy(ImVec2(size, size));   // Placeholder

    _DrawArc(size, max_angle_factor, percentage, thickness, pos);
}

File: tests/fixtures/ported/arc_progress_bar.py
"""arc_progress_bar: auto-ported from arc_progress_bar.hpp, arc_progress_bar.cpp
(upstream, MIT) by tools/autoport.py.

A mechanical port of Dear ImGui C++ to cmtk. Lines flagged
TODO(autoport) need a hand; everything else is the C++ under the
naming rules in tools/autoport.py.
"""
import math

import cmtk.im as im
def _get_style_color(color_id):
    """_GetStyleColor()."""
    assert(color_id >= 0  and  color_id < im.Col.COUNT)
    if color_id < 0  or  color_id >= im.Col.COUNT:
        return DEFAULT_FOREGROUND_COLOR
    return im.get_style().colors[color_id]

def _draw_arc(size, max_angle_factor, percentage, thickness, pos):
    """_DrawArc()."""
    draw_list = im.get_window_draw_list()
    x = pos[0]
    y = pos[1]
    ONE_DIV_360f = 1.0 / 360.0
    a_min_factor = 0.0
    a_max_factor = 0.0
    a_max_factor_100percentage = 0.0
    a_min_factor = -1.5 + ((360 - max_angle_factor) * ONE_DIV_360f)
    a_max_factor_100percentage = (a_min_factor + 1.0) * -1.0
    a_factor_delta = (a_max_factor_100percentage - a_min_factor) * (percentage * 0.01)
    a_max_factor = a_min_factor + a_factor_delta
    draw_list.path_arc_to((x + size * 0.5, y + size * 0.5), size * 0.5, 3.141592 * a_min_factor, 3.141592 * a_max_factor_100percentage)
    draw_list.path_stroke(_get_style_color(im.Col.BUTTON), im.DrawFlags.NONE, thickness)
    draw_list.path_arc_to((x + size * 0.5, y + size * 0.5), size * 0.5, 3.141592 * a_min_factor, 3.141592 * a_max_factor)
    draw_list.path_stroke(_get_style_color(im.Col.BUTTON_ACTIVE), im.DrawFlags.NONE, thickness)

def progress_bar_arc(size, max_angle_factor, percentage, pos, thickness):
    """ImGuiExt::ProgressBarArc()."""
    window_pos = im.get_window_pos()
    _draw_arc(size, max_angle_factor, percentage, thickness, (pos[0] + window_pos[0], pos[1] + window_pos[1]))

def progress_bar_arc(size, max_angle_factor, percentage, thickness):
    """ImGuiExt::ProgressBarArc()."""
    pos = im.get_cursor_screen_pos()
    im.dummy(size, size)
    _draw_arc(size, max_angle_factor, percentage, thickness, pos)
